### ISSUE TITLE
Walktrap misbehaves with disconnected graphs

### ISSUE THREAD

**Issue by szhorvat** [igraph/python-igraph#616]
`community_walktrap()` misbehaves with disconnected graphs:

```
In [18]: g=ig.Graph.Formula('1-2-3-4-1, 5-6-7-5')

In [19]: cl=g.community_walktrap()

In [20]: cl.as_clustering()
---------------------------------------------------------------------------
InternalError                             Traceback (most recent call last)
Cell In [20], line 1
----> 1 cl.as_clustering()

File ~/opt/miniconda3/envs/igdev2/lib/python3.10/site-packages/igraph-0.10.2-py3.10-macosx-10.9-x86_64.egg/igraph/clustering.py:832, in VertexDendrogram.as_clustering(self, n)
    830 num_elts = self._graph.vcount()
    831 idgen = UniqueIdGenerator()
--> 832 membership = community_to_membership(self._merges, num_elts, num_elts - n)
    833 membership = [idgen[m] for m in membership]
    834 return VertexClustering(
    835     self._graph, membership, modularity_params=self._modularity_params
    836 )

InternalError: Error at src/community/community_misc.c:94: Number of steps is greater than number of rows in merges matrix: found 6 steps, 5 rows. -- Invalid value
```

I fixed [a related issue in the C core](https://github.com/igraph/igraph/issues/2042) a while ago, but this one is specific to py-graph. Does it have to do with how the `optimal_count` value is set to 1 instead of actually finding the max modularity? I don't quite understand this conditional and when it takes the `False` branch:

https://github.com/igraph/python-igraph/blob/master/src/igraph/community.py#L389

UPDATE: I didn't realize that `[]` was treated as false in Python.

In short, I believe that the issue is that for disconnected graphs there are fewer merges than `vcount - 1`, and py-igraph does not take this into account like the C core does.

**Comment by szhorvat** [comment]
This line here:

https://github.com/igraph/python-igraph/blob/master/src/igraph/clustering.py#L830

should probably be the number of merges + 1 instead of the vcount.

**Comment by szhorvat** [comment]
The issue is that the number of components needs to be taken into account when calculating the `optimal_count` values. The problem is that `optimal_count` is calculated in many places. In general, when we have something like this:

```python
    if qs:
        optimal_count = qs.index(max(qs)) + 1
    else:
        optimal_count = 1
```

it should be instead:

```
    if qs:
        optimal_count = qs.index(max(qs))
    else:
        optimal_count = 0
    optimal_count += number_of_connected_components
```

This appears not just in walktrap but in all functions that produce dendrograms, as well as in the `optimal_count()` method of `VertexDendrogram`:

https://github.com/igraph/python-igraph/blob/master/src/igraph/clustering.py#L850

**Comment by ntamas** [comment]
It is a manifestation / duplicate of #17, the root cause is that incomplete dendrograms are not handled yet by the `VertexDendrogram` class. (The class assumes that the dendrogram is complete). An easy hack is to "fix" the dendrogram by connecting the components in arbitrary order, but I'm not quite satisfied with this solution.

**Comment by szhorvat** [comment]
Isn't it sufficient though to just do this in community finding functions?

```
    no_of_comps = graph.vcount() - len(merges)
    qs.reverse()
    optimal_count = qs.index(max(qs)) + no_of_comps
```

Checking `qs` for being empty isn't even necessary because the `modularity` vector returned by the C core should never be of zero length, not even for the null graph. The last element of the modularity vector will give the modularity when no partitioning was done at all. Thus there is always at least one element. This was inconsistent between functions in the past, but I fixed it a while ago.

Additionally, the same thing has to be handled when calculating the optimal number of clusters within `VertexDendrogram`.

There are various kinds of issues in different functions. For example, in the greedy modularity optimization we have this:

https://github.com/igraph/python-igraph/blob/master/src/igraph/community.py#L28

```python
    # qs may be shorter than |V|-1 if we are left with a few separated
    # communities in the end; take this into account
    diff = graph.vcount() - len(qs)
    qs.reverse()
    if qs:
        optimal_count = qs.index(max(qs)) + diff + 1
    else:
        optimal_count = diff
```

This is wrong because the `else` branch should have had a `diff+1` as well.  It would be cleaner to check `len(merges)` instead of `len(qs)`.

Note that we need to be careful with null graphs, which are best interpreted as having a single zero-size community (and not as having zero communities!!)

**Comment by ntamas** [comment]
> Isn't it sufficient though to just do this in community finding functions?

Well, yes, in theory it would be sufficient, but somehow I feel that this is just a band-aid that hides a deeper underlying problem, namely that `VertexDendrogram` cannot handle "partial" dendrograms where not all the nodes are merged all the way to a single top node, and my plan was always to fix that properly. I'll give it another go this time :)

**Comment by szhorvat** [comment]
Is there any part of `VertexDendrogram` that does not handle them well, other than the `optimal_count` method/property? Is there any other method of this class that misbehaves?

Or can we say that it works perfectly well (after fixing the `optimal_count` computation), however, the constructor has an inconvenient signature? It requires the user to specify the optimal partition count, which is inconvenient/unpleasant to compute. It would be more convenient if the offsetting of the index of of `qs`'s maximum could be done inside the `VertexDendrogram` constructor, so it wouldn't have to be repeated in basically every community finding function.

Note that I have not thought through the situation where the dendrogram information is given "from the other direction", i.e. splits instead of merges, as in the leading eigenvector function.

**Comment by ntamas** [comment]
> can we say that it works perfectly well (after fixing the optimal_count computation), however, the constructor has an inconvenient signature? It requires the user to specify the optimal partition count

No, it does not require it; `optimal_count` may be left at `None` (which is the default value), in which case the optimal count will be calculated on-the-fly when needed. This is what most users want. The reason why this parameter even exists is because igraph's C core often provides a merge matrix _and_ a sequence of modularities along the merges, and in that case we want to give a hint to the `VertexClustering` object so it does not have to recalculate the modularities internally again to determine the optimal count.

I've refactored the part that calculates the optimal count from the graph, the merge matrix and the modularity scores into a separate function.
